chia-client/src/peer.rs

```diff
@@ -127,18 +225,46 @@ impl Peer {
         Ok(())
     }
 
-    pub async fn request<T, R>(&self, body: T) -> Result<R>
+    pub async fn request<Response, Rejection, T>(&self, body: T) -> Result<Response, Rejection>
+    where
+        Response: Streamable + ChiaProtocolMessage,
+        Rejection: Streamable + ChiaProtocolMessage,
+        T: Streamable + ChiaProtocolMessage,
+    {
+        let message = self.request_raw(body).await?;
+        let data = message.data.as_ref();
+
+        if message.msg_type == Response::msg_type() {
+            Response::from_bytes(data).or(Err(Error::InvalidResponse(message)))
+        } else if message.msg_type == Rejection::msg_type() {
+            let rejection = Rejection::from_bytes(data).or(Err(Error::InvalidResponse(message)))?;
+            Err(Error::Rejection(rejection))
+        } else {
+            Err(Error::InvalidResponse(message))
+        }
+    }
+
+    pub async fn request_infallible<Response, T>(&self, body: T) -> Result<Response, ()>
+    where
+        Response: Streamable + ChiaProtocolMessage,
+        T: Streamable + ChiaProtocolMessage,
+    {
+        let message = self.request_raw(body).await?;
```

Comment:
in this case, the error returned by `?`, is that part of the co-routine as well?

Reply:
Not sure what you mean, if there is no response message because something fails, it will return the error early.